tests/ui/rfcs/rfc-0000-never_patterns/macros.rs

```diff
@@ -0,0 +1,40 @@
+// run-pass
+#![feature(never_patterns)]
+#![allow(incomplete_features)]
+
+use ParsedAs::*;
+#[derive(Debug, PartialEq, Eq)]
+enum ParsedAs {
+    Pattern,
+    Never,
+    Other,
+}
+
+macro_rules! detect_pat {
+    ($p:pat) => {
+        Pattern
+    };
+    (!) => {
+        Never
+    };
+    ($($x:tt)*) => {
+        Other
+    };
+}
+
+fn main() {
+    // For backwards compatibility this does not match `$p:pat`. This is similar to how or-patterns
+    // without parens don't match `$p:pat` either.
+    assert_eq!(detect_pat!(!), Never);
```

Comment:
Alternatively, you could make this not a runtime test but a compile-only test if you split the macro up and used `compile_fail!` on the macro branches you expect not to be hit.

e.g.

```
macro_rules! fail_if_pattern {
  ($pat:pat) => { compile_fail!(); }
  (!) => {}
}

fail_if_pattern!(!);
```

Reply:
I made it compile-time ina different way